Clean up bootstrap path handling

stuartwdouglas opened the pull request.

stuartwdouglas committed:
Clean up bootstrap path handling

aloubyansky commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/AdditionalApplicationArchiveBuildItem.java:
```diff
@@ -1,21 +1,20 @@
     }
 
-    public Path getPath() {
+    public PathsCollection getPath() {
```
This may break external extensions.

stuartwdouglas replied:
I changed it to getPaths and added a deprecated method

aloubyansky reviewed (commented):
I added a few questions.

stuartwdouglas commented on devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusGenerateConfig.java:
```diff
@@ -57,6 +57,7 @@ public void buildQuarkus() {
                 .setTargetDirectory(getProject().getBuildDir().toPath())
                 .setBaseName(extension().finalName())
                 .setAppArtifact(appArtifact)
+                .setApplicationRoot(appArtifact.getPaths())
```
I just think it is better to be super explicit about where everything comes from. Before it was kinda implicit, but not everywhere was using the same logic.

stuartwdouglas commented on devtools/maven/src/main/java/io/quarkus/maven/GenerateConfigMojo.java:
```diff
@@ -132,6 +132,8 @@ public void execute() throws MojoExecutionException {
                     .builder()
                     .setAppArtifact(appArtifact)
+                    .setApplicationRoot(classesDir)
+                    .setProjectRoot(project.getBasedir().toPath())
```
Yes, tests create their own tmp dir which is the app root, but uses metadata from the project root.

aloubyansky replied:
I meant in this specific case. I don't think all these have to required. It does not look simpler to me.

stuartwdouglas replied:
I want it to be explicit exactly which paths are used where. At the moment if one is null it just picks a different one. We only have a few different places that use this launcher code, IMHO it is better to be explicit rather than rely on implicit behaviour.

stuartwdouglas commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/QuarkusBootstrap.java:
```diff
@@ -314,6 +314,10 @@ public Builder setDependenciesOrigin(DependenciesOrigin dependenciesOrigin) {
+        /**
+         * The app artifact. Note that if you want to use this as the basis of the application
+         * you must also explicitly set the application root to this artifacts paths.
+         */
```
TBH I don't know if we event should keep the app artifact concept. App roots == classes, project root == dependency metadata, so I don't know if we can remove this concept and just add anything else it provides as explicit metadata.

aloubyansky replied:
There is a special case. In the platform we import the test JARs, in which case we need the app artifact to resolve its dependencies aligning them with the platform BOM.

aloubyansky replied:
Here is what I'm thinking. We shouldn't be setting projectRoot at all anywhere. The refactoring I previously did introduced BootstrapMavenContext which figures it out anyway if workspace discovery is enabled and it's necessary. So I think unless a test creates a custom workspace at a tmp location, I don't think we need to be setting projectRoot explicitly. I can check whether it breaks somewhere.
With the projectRoot out, we are left with the classes and resources output dirs and the app artifact. Which is already a lot better.
When we run the platform TS or re-building the app from the JAR (version updates, etc, which is not yet a real use-case but prototyped) all that will be coming from the app artifact. My preference would be to have the app artifact as the default source of the app classes and resources. Wit the possibility to override.
I agree that we shouldn't not be probably treating the app root artifact differently than the other user deps. But I think it'll require some work.
If you like I can look into this after #9099 has been merged. I think it's ready now.

stuartwdouglas replied:
We have 3 different uses that do create code in a tmp dir, so we definitely still need the concept. 

What do we actually take from the app artifact? We use its classes and resources, but this is covered by the application root. If the only other thing that it provides is dependency metadata then this is basically similar to the project root. 

In a lot of our cases (e.g. tests, dev mode) there is no app artifact, so IMHO it should not be the default.

Maybe you are allowed to either call setApplicationRoot()and setProjectRoot() or you can call setAppArtifact(), but not both in the same bootstrap? 

I just want it to be completely unambiguous where everything comes from.

aloubyansky replied:
> Maybe you are allowed to either call setApplicationRoot()and setProjectRoot() or you can call setAppArtifact(), but not both in the same bootstrap?

I agree with the direction this is going. Setting the project root should be redundant though now, as i previosly mentioned. Although you if you prefer setting it explicitly, that's not a problem.
There is still a special case though when we integrate external test jars into the platform. In this case there is an app artifact but we still use the context of the platform project to manage the dependency versions of the test app and other maven config. Although that could probably be viewed as an empty jar pom-only project with a dep on the app to be tested. In fact, that's what it is. So, I guess, it should work.

aloubyansky replied:
I meant the empty jar frpm the traditional maven project config perspective. We'd still have to build an executable for it.

stuartwdouglas committed:
Add support for live reload when launching from IDE

stuartwdouglas commented:
@aloubyansky I think I have addressed your concerns, assuming CI is green can we get this merged for now and then talk about the next steps?

This PR now also include proper support for right click to run for maven projects, which is a pretty nice feature.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java:
```diff
@@ -208,28 +208,20 @@ public synchronized QuarkusClassLoader getBaseRuntimeClassLoader() {
 
-                for (AdditionalDependency i : quarkusBootstrap.getAdditionalApplicationArchives()) {
-                    //src/test is the highest priority
-                    if (i.isTestClassRoot()) {
```
So, the test paths are now expected to be among the app root paths, correct?

stuartwdouglas replied:
Yes, there is no special case handling for them any more, they are just installed first so their application.properties takes precidence.

aloubyansky commented on independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/QuarkusBootstrap.java:
```diff
@@ -127,8 +125,7 @@ public CuratedApplication bootstrap() throws BootstrapException {
                 .setForcedDependencies(forcedDependencies)
-                .setProjectRoot(getProjectRoot() != null ? getProjectRoot()
-                        : getApplicationRoot());
+                .setProjectRoot(getProjectRoot());
```
I'll check whether this works for the platform TS. There is still a bit of a mess around these.

stuartwdouglas replied:
After this PR I want to get rid of the whole ProjectRoot thing and replace it with a more general MetadataSource concept. Getting IDE run to work in gradle is going to take some work, and this refactoring will be part of that. Lets discuss on zulip when you are around.

aloubyansky commented on test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java:
```diff
@@ -96,32 +96,30 @@ private ExtensionState doJavaStart(ExtensionContext context) throws Throwable {
+            runnerBuilder.setProjectRoot(Paths.get("").normalize().toAbsolutePath());
+
+            rootBuilder.add(appClassLocation);
+            runnerBuilder.setApplicationRoot(rootBuilder.build());
```
This needs to be checked whether it works for the platform TS.

aloubyansky reviewed (commented):
I added a few comments and I need to check whether it works for the platform TS.

aloubyansky commented:
Here is one breaking change
````
Caused by: java.lang.NoSuchMethodError: io.quarkus.deployment.builditem.ArchiveRootBuildItem.getArchiveLocation()Ljava/nio/file/Path;
	at org.kie.kogito.quarkus.deployment.KogitoAssetsProcessor.generateModel(KogitoAssetsProcessor.java:176)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:932)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
````

aloubyansky commented:
What I think we should do is have a run with all the deprecated methods remove, identify failures and create issues in external extension projects to adapt asap.

aloubyansky commented on core/deployment/src/main/java/io/quarkus/deployment/builditem/ArchiveRootBuildItem.java:
```diff
@@ -98,17 +89,6 @@ private ArchiveRootBuildItem(Builder builder, QuarkusBuildCloseablesBuildItem bu
-    public Path getArchiveLocation() {
-        return archiveLocation;
-    }
-
```
Here is a variant of the method that will keep the platform TS passing, given that the test paths are added first we are returning the last on the list
````
    /**
     * If this archive is a jar file it will return the path to the jar file on the file system,
     * otherwise it will return the directory that this corresponds to.
     *
     * @deprecated in favor of {@link #getPaths()}
     */
    @Deprecated
    public Path getArchiveLocation() {
        final Iterator<Path> i = paths.iterator();
        Path last = i.next();
        while (i.hasNext()) {
            last = i.next();
        }
        return last;
    }
````

aloubyansky replied:
The good thing is removing `public Path getArchiveRoot()` does not actually break the platform.

aloubyansky commented:
@mariofusco @evacchi could please make sure you don't use `ArchiveRootBuildItem.getArchiveLocation()`? It is deprecated and going to be removed in the near future. See the stacktrace in https://github.com/quarkusio/quarkus/pull/9139#issuecomment-627836968

aloubyansky commented:
I didn't find any other platform issue with this change. Perhaps, we could keep the deprecated methods until the next release, just in case?

evacchi commented:
@aloubyansky what's the correct new way to get the same result? IIRC we used it to get the path to the current module and/or `target` subdirectory

aloubyansky commented:
Yes, the method was deprecated already in 1.4.2.Final. Let's discuss it on zulip.
